compiler/rustc_passes/src/liveness.rs

```diff
@@ -1274,6 +1304,47 @@ impl<'a, 'tcx> Liveness<'a, 'tcx> {
 
         ln
     }
+
+    fn warn_about_unreachable(
+        &mut self,
+        orig_span: Span,
+        orig_ty: Ty<'tcx>,
+        expr_span: Span,
+        expr_id: HirId,
+        descr: &str,
+    ) {
+        if !orig_ty.is_never() {
+            // Unreachable code warnings are already emitted during type checking.
+            // However, during type checking, full type information is being
+            // calculated but not yet available, so the check for diverging
+            // expressions due to uninhabited result types is pretty crude and
+            // only checks whether ty.is_never(). Here, we have full type
+            // information available and can issue warnings for less obviously
+            // uninhabited types (e.g. empty enums). The check above is used so
+            // that we do not emit the same warning twice if the uninhabited type
+            // is indeed `!`.
```

Comment:
Is there a reason _not_ to remove the current, existing, crude check in favor of this one?

Reply:
Yes, because type inference makes use of the crude divergence information available, e.g. here:
https://github.com/rust-lang/rust/blob/5dc8789e300930751a78996da0fa906be5a344a2/compiler/rustc_typeck/src/check/fn_ctxt/checks.rs#L654-L660
The crude check is here:
https://github.com/rust-lang/rust/blob/5dc8789e300930751a78996da0fa906be5a344a2/compiler/rustc_typeck/src/check/expr.rs#L202-L205
Unfortunately, we can't replace it with `tcx.is_ty_uninhabited_from()` because type information has not been committed to `tcx` yet at this point.